### PR TITLE
Docs + Doctest

### PR DESCRIPTION
**Pull request by orausch**
Mainly ported over the existing documentation from the readme into sphinx.

- autogenerate documentation for all supported ONNX nodes
- add examples and test them with doctest

TODO:
- [x] make readme long again
- [x] ``rg [^`]`[^`]``
- [x] move setup code to before doctest
- [x] remove ORT install from RTD build

### CODE REVIEW

**Comment by orausch** [comment]
@tbennun I don't have perms to add readthedocs to this repo. We will need to still do that (and clean up the README) before merging.

**Comment by orausch** [comment]
Built docs: https://daceml.readthedocs.io/en/docs/

File: README.md [hunk 1]
```diff
@@ -1,130 +1,11 @@
 [![Build Status](https://travis-ci.org/spcl/daceml.svg?branch=master)](https://travis-ci.org/spcl/daceml)
 [![codecov](https://codecov.io/gh/spcl/daceml/branch/master/graph/badge.svg)](https://codecov.io/gh/spcl/daceml)
-# daceml
-Machine learning powered by data-centric parallel programming. 
+[![Documentation Status](https://readthedocs.org/projects/daceml/badge/?version=latest)](https://daceml.readthedocs.io/en/latest/?badge=latest)
 
-This project adds PyTorch and ONNX model loading support to [DaCe](https://github.com/spcl/dace), and supports ONNX operators as library nodes to DaCe stateful dataflow multigraphs.
```
Comment (tbennun) [inline README.md:6]: README is way too empty. You need some teasers of features and links to the appropriate doc pages.

File: daceml/onnx/nodes/onnx_op.py [hunk 2]
```diff
@@ -52,12 +94,16 @@ class ONNXOp(nd.LibraryNode):
     def iter_outputs_in_onnx_order(self, state):
```
Comment (tbennun) [inline daceml/onnx/nodes/onnx_op.py:94]: would be nice to have type hints here and below. There's also a return value that's missing from the doc.

File: daceml/onnx/nodes/onnx_op.py [hunk 3]
```diff
@@ -427,11 +475,42 @@ def __init__(self, name, *args, location=None, **op_attributes):
...
 
-    attrs['__init__'] = __init__
+    input_connector_docstrings = "\n".join(
+        _get_connector_docstring(param) for param in dace_schema.inputs)
+    output_connector_docstrings = "\n".join(
+        _get_connector_docstring(param) for param in dace_schema.outputs)
 
     cls_name = "ONNX" + dace_schema.name
+
+    # the first line of the init docstring contains the signature of the method. This will be picked up by sphinx so
```
Comment (tbennun) [inline daceml/onnx/nodes/onnx_op.py:485]: comment line too long?

File: daceml/pytorch/module.py [hunk 4]
```diff
@@ -13,24 +14,50 @@
...
+
+            >>> from daceml.pytorch import DaceModule
+            >>> class MyModule(nn.Module):
+            ...     def forward(self, x):
+            ...        x = torch.log(x)
+            ...        x = torch.sqrt(x)
+            ...        return x
+            >>> module = MyModule()
+            >>> module(torch.ones(2))
+            tensor([0., 0.])
```
Comment (tbennun) [inline daceml/pytorch/module.py:36]: this seems like the wrong answer! :(
same for example below

File: daceml/pytorch/module.py [hunk 5]
```diff
@@ -55,16 +83,31 @@ def initialize_sdfg(self, dummy_inputs) -> ONNXModel:
...
+            ... class MyModule(nn.Module):
+            ...     def forward(self, x):
+            ...        x = torch.log(x)
+            ...        x = torch.sqrt(x)
+            ...        return x
+            >>> module = MyModule()
+            >>> module(torch.ones(2))
+            Automatically expanded library node "ONNX_Log_0".
+            Automatically expanded library node "ONNX_Sqrt_1".
+            array([0., 0.], dtype=float32)
```
Comment (tbennun) [inline daceml/pytorch/module.py:110]: same

File: daceml/pytorch/module.py [hunk 6]
```diff
@@ -13,24 +14,50 @@
...
+            ...        x = torch.sqrt(x)
+            ...        return x
+            >>> module = MyModule()
+            >>> module(torch.ones(2))
+            tensor([0., 0.])
+            >>> dace_module = DaceModule(module)
+            >>> dace_module(torch.ones(2))
+            Automatically expanded library node "ONNX_Log_0".
+            Automatically expanded library node "ONNX_Sqrt_1".
+            array([0., 0.], dtype=float32)
```
Comment (tbennun) [inline daceml/pytorch/module.py:41]: same

File: daceml/pytorch/module.py [hunk 7]
```diff
@@ -13,24 +14,50 @@
...
+            ...        x = torch.log(x)
+            ...        x = torch.sqrt(x)
+            ...        return x
+            >>> module = MyModule()
+            >>> module(torch.ones(2))
+            tensor([0., 0.])
+            >>> dace_module = DaceModule(module)
+            >>> dace_module(torch.ones(2))
+            Automatically expanded library node "ONNX_Log_0".
+            Automatically expanded library node "ONNX_Sqrt_1".
```
Comment (tbennun) [inline daceml/pytorch/module.py:40]: why does torch.ones need log and sqrt? :)

File: doc/conf.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+import subprocess
+model_path = os.path.join("..", "tests", "onnx_files", "efficientnet.onnx")
+# Download model
+if not os.path.exists(model_path):
+    subprocess.check_call([
+        "wget",
+        "https://github.com/onnx/models/raw/master/vision/classification/efficientnet-lite4/model/efficientnet-lite4-11.onnx",
+        "--output-document={}".format(model_path)
+    ])
+'''
```
Comment (tbennun) [inline doc/conf.py:65]: why do we want this here?

**Review by tbennun — CHANGES_REQUESTED** [review]
Looks good, other than some comments about the docs.

File: doc/conf.py (hunk 8 above)
Comment (orausch) [reply doc/conf.py:65]: This is the setup code for the doctest. One of the examples is efficientnet, so we just download it if it's not there already (it should be there anyway if pytest ran before this)

File: daceml/onnx/schema.py [hunk 9]
```diff
@@ -129,14 +158,14 @@ def __repr__(self):
...
-    Int = ()
-    Float = ()
-    String = ()
-    Ints = ()
-    Floats = ()
-    Strings = ()
-    Tensor = ()
-    Unsupported = ()
+    Int = ()  #: Integer (python representation is ``int``)
+    Float = ()  #: Float (python representation is ``double``)
```
Comment (tbennun) [inline daceml/onnx/schema.py:162]: ???

File: daceml/onnx/schema.py [hunk 10]
```diff
@@ -129,14 +158,14 @@ def __repr__(self):
...
+    String = ()  #: String (python representation is ``str``)
+    Ints = ()  #: Ints (python representation is ``List`` [``int``])
+    Floats = ()  #: Floats (python representation is ``List`` [``double``])
```
Comment (tbennun) [inline daceml/onnx/schema.py:165]: shouldn't it be float?

**Review by tbennun — APPROVED** [review]
Note minor change!

File: daceml/onnx/schema.py (hunk 10 above)
Comment (orausch) [reply daceml/onnx/schema.py]: yup

**Pull request merged**